sbi/diagnostics/sbc.py

```diff
@@ -0,0 +1,221 @@
+        # to update the pbar only after the workers finished a task.
+        with tqdm_joblib(
+            tqdm(
+                thos_batches,
+                disable=not show_progress_bars,
+                desc=f"""Running {num_sbc_samples} sbc runs in {len(thos_batches)}
+                    batches.""",
+                total=len(thos_batches),
+            )
+        ) as progress_bar:
+            sbc_outputs = Parallel(n_jobs=num_workers)(
+                delayed(sbc_on_batch)(
+                    thos_batch, xos_batch, posterior, num_posterior_samples, ranking_rv
+                )
+                for thos_batch, xos_batch in zip(thos_batches, xos_batches)
+            )
+    else:
+        pbar = tqdm(
+            total=num_sbc_samples,
+            disable=not show_progress_bars,
+            desc=f"Running {num_sbc_samples} sbc samples.",
+        )
+
+        with pbar:
+            sbc_outputs = []
+            for thos_batch, xos_batch in zip(thos_batches, xos_batches):
+                sbc_outputs.append(
+                    sbc_on_batch(
+                        thos_batch,
+                        xos_batch,
+                        posterior,
+                        num_posterior_samples,
+                        ranking_rv,
+                    )
+                )
+                pbar.update(sbc_batch_size)
+
+    # Aggregate results.
+    ranks = []
+    dap_samples = []
+    log_probs = []
+    for out in sbc_outputs:
+        ranks.append(out[0])
+        log_probs.append(out[1])
+        dap_samples.append(out[2])
+
+    ranks = torch.cat(ranks)
+    dap_samples = torch.cat(dap_samples)
+    log_probs = torch.cat(log_probs)
+
+    return ranks, log_probs, dap_samples
+
+
+def sbc_on_batch(thos, xos, posterior, L, ranking_rv):
+    """Return SBC results for a batch of SBC parameters and data from prior.
+
+    Returns
+        ranks: ranks of true parameters vs. posterior samples under the specified RV,
+            for each posterior dimension.
+        log_prob_thos: log prob of true parameters under the approximate posterior.
+        dap_samples: samples from the data averaged posterior for the current batch,
+            i.e., a single sample from each approximate posterior.
+    """
+
+    log_prob_thos = torch.zeros(thos.shape[0])
+    dap_samples = torch.zeros_like(thos)
+    ranks = torch.zeros_like(thos)
+
+    for idx, (tho, xo) in enumerate(zip(thos, xos)):
+        # Log prob of true params under posterior.
+        log_prob_thos[idx] = posterior.log_prob(tho, x=xo)
+
+        # Rank true params vs posterior params under ranking_rv.
+        ths = posterior.sample((L,), x=xo, show_progress_bars=False)
+        dap_samples[idx] = ths[
+            0,
+        ]
+        # rank for each posterior dimension.
+        for dim in range(thos.shape[1]):
+            ranks[idx, dim] = (
+                (ranking_rv.log_prob(ths[:, dim]) < ranking_rv.log_prob(tho[dim]))
```

Comment:
I am not sure why you need to `ranking_rv.log_prob` here? Can you elaborate? I understood the paper in a sense that `ranking_rv(theta)` is enough?